### ISSUE TITLE
C++ API: unable to find cross entropy loss

### ISSUE THREAD

**Issue by ejoebstl** [pytorch/pytorch#16696]
I am unable to find the cross entropy loss function for the C++ API. The C++ API docs show zero results when looking for it. 

Also, it's not `torch::cross_entropy` or `torch::cross_entropy_loss`.

cc @yf225 @glaringlee

**Comment by zou3519** [comment]
Does it exist? cc @yf225

**Comment by yf225** [comment]
@ejoebstl the `torch::cross_entropy` module is not implemented in the C++ API yet (we plan to add it soon). We can use the following to achieve it now:

```cpp
// Based on https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1998
torch::nll_loss(torch::log_softmax(input, 1), target, weight, reduction, ignore_index)
```

The accepted values for `reduction` are in aten/src/ATen/core/Reduction.h

**Comment by sidazhang** [comment]
> ```c++
> // Based on https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1998
> torch::nll_loss(torch::log_softmax(input, 1), target, weight, reduction, ignore_index)
> ```

The python version is a fused operator though for better numerical stability. Am I right?

**Comment by apaszke** [comment]
I don't think so. IIRC most of the instability comes from using softmax and is alleviated by log_softmax.

**Comment by ezyang** [comment]
cross_entropy exists in torch/csrc/api/include/torch/nn/functional/loss.h now so I think we're good.
